### PR TITLE
fix ExportsInfo

### LINKED ISSUE #15214: Incorrect cache invalidation using module with side effects and re-exporting from empty file
# Bug report

Cache get invalidated on the second run when using a module with side effects (like `console.log`) and re-exporting from an empty file.

**What is the current behavior?**
The cache pack gets invalid on the second run. On the third run and more cache hits.

**If the current behavior is a bug, please provide the steps to reproduce.**

https://github.com/7rulnik/webpack-repro-empty-export-cache-invalidation

There is side effect https://github.com/7rulnik/webpack-repro-empty-export-cache-invalidation/blob/main/convert/typescript/es.object.assign.js#L1
And there re-export from empty file https://github.com/7rulnik/webpack-repro-empty-export-cache-invalidation/blob/main/convert/typescript/index.js#L2

Run `yarn webpack` twice. You will find a log like this:

```
[webpack.cache.PackFileCacheStrategy] Pack got invalid because of write to:
ResolverCachePlugin|normal|dependencyType=|esm|path=|/Users/v7rulnik/projects/webpack-cache-test|request=|./index.js
```

**What is the expected behavior?**

A cache should not get invalidated on the second run.

**Other relevant information:**

You can come to the question *«Why re-export from an empty file? Just delete it»*. 
Repro is based on the real case of Storybook. Storybook imports `@storybook/addon-docs/dist/esm/frameworks/react/config.js`.

Source code of addon-docs looks like this
https://github.com/storybookjs/storybook/blob/6657698254b26df6aa7c4143eae8593bbfe5c208/addons/docs/src/lib/convert/typescript/index.ts. But `types.ts` contains only types https://github.com/storybookjs/storybook/blob/6657698254b26df6aa7c4143eae8593bbfe5c208/addons/docs/src/lib/convert/typescript/types.ts — this is why we get empty file. You can check published file https://unpkg.com/browse/@storybook/addon-docs@6.4.13/dist/esm/lib/convert/typescript/types.js And in this case, side effects come from `core-js` https://unpkg.com/browse/@storybook/addon-docs@6.4.13/dist/esm/lib/convert/typescript/convert.js. See https://unpkg.com/core-js@3.20.3/modules/es.array.map.js.

Also, the bug doesn't reproduce when `usedExports` is set to false.

----------------

webpack version:  5.66.0
Node.js version: 14.17.6
Operating System: MacOS 12.0.1 (21A559) ARM

### PR DESCRIPTION
**Pull request by vankop**
- should mark every export `canMangleProvide=false` if there are unknown exports
- add `__webpack_exports_info__.<name>.canMangle` to api
- add test case

**What kind of change does this PR introduce?**

bugfix
fixes #15214 

**Did you add tests for your changes?**

yes

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

add `__webpack_exports_info__.<name>.canMangle` to api

### CODE REVIEW

File: test/__snapshots__/StatsTestCases.basictest.js.snap [hunk 1]
```diff
@@ -3274,7 +3274,7 @@ cacheable modules 1.22 KiB
     ModuleConcatenation bailout: Module is not an ECMAScript module
 webpack x.x.x compiled successfully in X ms
 
-asset main.no-side.js 979 bytes [emitted] [minimized] (name: main)
+asset main.no-side.js 993 bytes [emitted] [minimized] (name: main)
```
Comment (vankop) [inline test/__snapshots__/StatsTestCases.basictest.js.snap:3277]: no mangling for one variable

File: lib/dependencies/ExportsInfoDependency.js [hunk 2]
```diff
@@ -46,6 +46,12 @@ const getProperty = (moduleGraph, module, exportName, property, runtime) => {
 		}
 	}
 	switch (property) {
+		case "canMangle": {
+			const exportsInfo = moduleGraph.getExportsInfo(module);
+			const exportInfo = exportsInfo.getExportInfo(exportName);
+			if (exportInfo) return exportInfo.canMangle;
+			return exportsInfo.otherExportsInfo.canMangleProvide;
```
Comment (sokra) [inline lib/dependencies/ExportsInfoDependency.js:53]: Why is that `canMangleProvide` and not `canMangle`?

```suggestion
			return exportsInfo.otherExportsInfo.canMangle;
```

File: test/configCases/parsing/harmony-reexport/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
+/** @type {import("../../../../").Configuration} */
+module.exports = {};
```
Comment (sokra) [inline test/configCases/parsing/harmony-reexport/webpack.config.js:2]: When this doesn't need special config, we should move it into `test/cases` so it's tested with different configs.

Maybe wrapping some assertions with `if(process.env.NODE_ENV === "production")` (e. g. for optimizations that only run in production mode)

**Review by sokra — CHANGES_REQUESTED** [review]

File: test/configCases/parsing/harmony-reexport/webpack.config.js (hunk 3 above)
Comment (vankop) [reply test/configCases/parsing/harmony-reexport/webpack.config.js:2]: this was done intentionally to run 3 times (check for cache invalidation)

Comment (sokra) [reply test/configCases/parsing/harmony-reexport/webpack.config.js:2]: Ah yes that makes sense

File: test/TestCasesCachePack.longtest.js [hunk 4]
```diff
@@ -1,9 +1,44 @@
...
+					category: "json",
+					test: "import-assertions-type-json"
+				},
+				{
+					// Pack got invalid because of write to: Compilation/modules|/Users/ivankopeykin/Repositories/webpack/test/cases/loaders/no-string/loader.js!/Users/ivankopeykin/Repositories/webpack/test/cases/loaders/no-string/file.js
+					category: "loaders",
+					test: "no-string"
+				},
+				{
+					// Pack got invalid because of write to: ResolverCachePlugin|normal|dependencyType=|esm|path=|/Users/ivankopeykin/Repositories/webpack/test/cases|request=|./large/big-assets/
```
Comment (vankop) [inline test/TestCasesCachePack.longtest.js:30]: fix in https://github.com/webpack/webpack/pull/15373

File: test/TestCasesCachePack.longtest.js [hunk 5]
```diff
@@ -4,6 +4,37 @@ const { describeCases } = require("./TestCases.template");
 describe("TestCases", () => {
 	describeCases({
 		name: "cache pack",
+		infrastructureLogErrors: {
+			compile: {
+				// Module build failed
+				["error-hide-stack"]:
```
Comment (vankop) [inline test/TestCasesCachePack.longtest.js:10]: maybe add this to `errors.js` file?

File: test/ConfigCacheTestCases.longtest.js [hunk 6]
```diff
@@ -1,41 +1,35 @@
...
-			{
-				// Pack got invalid because of write to: webpack.HttpUriPlugin|https://raw.githubusercontent.com//webpack//webpack//main/CODE_OF_CONDUCT.md
-				category: "asset-modules",
-				test: "http-url"
-			}
-		],
-		filter: [logErrors.PERSISTENCE_CACHE_INVALIDATE_ERROR]
+		wasm: {
+			// Can not compile wasm module
+			["missing-wasm-experiment"]:
```
Comment (sokra) [inline test/ConfigCacheTestCases.longtest.js:8]:
```suggestion
			"missing-wasm-experiment":
```

That's not necessary... but you can leave that for now

**Comment by sokra** [comment]
Thanks

**Pull request merged**